### PR TITLE
IDEMPIERE-5866: different identifiers of info window and m_table cause in…

### PR DESCRIPTION
**Pull request by hieplq**
…fo window fill query text to wrong identifier column

https://idempiere.atlassian.net/browse/IDEMPIERE-5866

# Pull Request Checklist

- [x] My code follows the [code guidelines](https://wiki.idempiere.org/en/Contributing_to_Trunk) of this project
- [x] My code follows the best practices of this project
- [x] I have performed a self-review of my own code
- [x] My code is easy to understand and review. 
- [x] I have checked my code and corrected any misspellings
- [x] In hard-to-understand areas, I have commented my code.
- [x] My changes generate no new warnings
### Tests
- [ ] I have tested the direct scenario that my code is solving
- [ ] I checked all collaterals that can be affected by my changes, and tested other potential affected scenarios
- [ ] New and existing unit tests pass locally with my changes
- [ ] I have added unit tests that prove my fix is effective or that my feature works
### Documentation
- [ ] I have made corresponding changes to the documentation as follows:
- - [ ] New feature (non-breaking change which adds functionality): I have created the New Feature page in the project wiki explaining the functionality and how to use it. If relevant, I have committed sample data to the core seed to have usable examples in GardenWorld.
- - [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected): I have documented the change in a clear way that everyone in the community can understand the impact of the change.
- - [ ] Improvement (improves and existing functionality): This documentation is needed if the improvement changes the way the user interacts with the system or the outcome of a process/task changes. If it is just, for instance, a performance improvement, documentation might not be needed. 
- [ ] The changed/added documentation is in the project wiki (not privately-hosted pdf files or links pointing to a company website) and is complete and self-explanatory.

### CODE REVIEW

File: org.adempiere.ui.zk/WEB-INF/src/org/adempiere/webui/info/InfoWindow.java [hunk 1]
```diff
@@ -646,11 +648,55 @@ protected void processQueryValue() {
...
 				String separator = MSysConfig.getValue(MSysConfig.IDENTIFIER_SEPARATOR, "_", Env.getAD_Client_ID(Env.getCtx()));
 				String[] values = queryValue.split("[" + separator.trim()+"]");
-				if (values.length == 2) {
+				if (values.length > 0) {
 					splitValue = true;
-					for(int i = 0; i < values.length && i < identifiers.size(); i++) {
-						WEditor editor = identifiers.get(i);
-						editor.setValue(values[i].trim());
+					
+					// store identifiers on info window already sort follow identify on m_table
```
Comment (hengsin) [inline org.adempiere.ui.zk/WEB-INF/src/org/adempiere/webui/info/InfoWindow.java:654]: // store identifiers on info window, sort to follow identifier on m_table

File: org.adempiere.ui.zk/WEB-INF/src/org/adempiere/webui/info/InfoWindow.java [hunk 2]
```diff
@@ -646,11 +648,55 @@ protected void processQueryValue() {
...
+					List<WEditor> fillIdentifiers = new ArrayList<>();
+					// store query value already ignore value for identify not exists on info window
```
Comment (hengsin) [inline org.adempiere.ui.zk/WEB-INF/src/org/adempiere/webui/info/InfoWindow.java:656]: // store query value, ignore value for identifier not exists on info window

File: org.adempiere.ui.zk/WEB-INF/src/org/adempiere/webui/info/InfoWindow.java [hunk 3]
```diff
@@ -646,11 +648,55 @@ protected void processQueryValue() {
...
+							tableIdentifiers = mLookup.getLookupInfo().lookupDisplayColumnNames;
+					}
+					
+					if (tableIdentifiers != null) {
+						for (int i = 0; i < tableIdentifiers.size(); i++) {
+							// final local variable to access inside lambda expression
+							int indexFinal = i;
+							List<String> tableIdentifiersFinal = tableIdentifiers;
+							
+							// sort identify of info window follow m_table
```
Comment (hengsin) [inline org.adempiere.ui.zk/WEB-INF/src/org/adempiere/webui/info/InfoWindow.java:675]: // sort identifiers of info window to follow m_table

File: org.adempiere.ui.zk/WEB-INF/src/org/adempiere/webui/info/InfoWindow.java [hunk 4]
```diff
@@ -646,11 +648,55 @@ protected void processQueryValue() {
...
+								if (identifierEditor.getColumnName().equals(tableIdentifiersFinal.get(indexFinal))) {
+									fillIdentifiers.add(identifierEditor);
+									fillValues.add(values[indexFinal]);
+								}
+							});
+						}
+					}
+					
+					// case not exists mLookup.getLookupInfo().lookupDisplayColumnNames
+					// or non identify on info window exists on m_table
```
Comment (hengsin) [inline org.adempiere.ui.zk/WEB-INF/src/org/adempiere/webui/info/InfoWindow.java:687]: // or no identifiers on info window exists on m_table

File: org.adempiere.ui.zk/WEB-INF/src/org/adempiere/webui/info/InfoWindow.java [hunk 5]
```diff
@@ -646,11 +648,55 @@ protected void processQueryValue() {
...
+									fillIdentifiers.add(identifierEditor);
+									fillValues.add(values[indexFinal]);
+								}
+							});
+						}
+					}
+					
+					// case not exists mLookup.getLookupInfo().lookupDisplayColumnNames
+					// or non identify on info window exists on m_table 
+					// fail back to old logic just set values to parameter
```
Comment (hengsin) [inline org.adempiere.ui.zk/WEB-INF/src/org/adempiere/webui/info/InfoWindow.java:688]: // fall back to old logic and just set values to parameter

File: org.adempiere.ui.zk/WEB-INF/src/org/adempiere/webui/info/InfoWindow.java [hunk 6]
```diff
@@ -646,11 +648,55 @@ protected void processQueryValue() {
...
+					if (fillIdentifiers.size() == 0) {
+						for(int i = 0; i < values.length && i < identifiers.size(); i++) {
+							fillIdentifiers.add(identifiers.get(i));
+							fillValues.add(values[i]);
+						}
+					}
+					
+					// do fill value to editor (both case correct order and non-correct order by fail back)
```
Comment (hengsin) [inline org.adempiere.ui.zk/WEB-INF/src/org/adempiere/webui/info/InfoWindow.java:696]: // do fill value to editor (for both corrected order and fall back)

**Comment by hieplq** [comment]
because IDEMPIERE-5864 base on IDEMPIERE-5866 so i push it here also

File: org.adempiere.ui.zk/WEB-INF/src/org/adempiere/webui/info/InfoWindow.java [hunk 7]
```diff
@@ -617,55 +619,131 @@ public void moveProcessButtonsToBeforeRight() {
...
+				log.log(Level.SEVERE, "error", ex.getCause());
+			}
+		}else {
+			// => don't run test in case not found auto complete column
+			log.log(Level.SEVERE, String.format("can't found column for autocomplete query for field %s - field id %s. first identify column on m_table need to exists on identifies of info window", 
+					m_gridfield.getColumnName(), m_gridfield.getAD_Column_ID()));
+		}
+	}
+	
+	protected void testQueryForEachIdentify() {
```
Comment (hengsin) [inline org.adempiere.ui.zk/WEB-INF/src/org/adempiere/webui/info/InfoWindow.java:676]: It should be "identifier" not "identify"

File: org.adempiere.ui.zk/WEB-INF/src/org/adempiere/webui/info/InfoWindow.java [hunk 8]
```diff
@@ -617,55 +619,131 @@ public void moveProcessButtonsToBeforeRight() {
...
 		}
 		
-		if (m_count > 0) {
-			executeQuery();
-			renderItems();
-		} else if (!splitValue) {
-			editors.get(0).setValue(queryValue);
+		// case not exists mLookup.getLookupInfo().lookupDisplayColumnNames
+		// or no identifiers on info window exists on m_table
+		// fail back to old logic just set values to parameter
```
Comment (hengsin) [inline org.adempiere.ui.zk/WEB-INF/src/org/adempiere/webui/info/InfoWindow.java:731]: // fall back to old logic and just set values to identifiers

File: org.adempiere.ui.zk/WEB-INF/src/org/adempiere/webui/info/InfoWindow.java [hunk 9]
```diff
@@ -617,55 +619,131 @@ public void moveProcessButtonsToBeforeRight() {
...
+		if (fillIdentifiers.size() == 0) {
+			for(int i = 0; i < values.length && i < identifiers.size(); i++) {
+				fillIdentifiers.add(identifiers.get(i));
+				fillValues.add(values[i]);
+			}
 		}
-		isQueryByUser = false;
+		
+		// do fill value to editor (both case correct order and non-correct order by fail back)
```
Comment (hengsin) [inline org.adempiere.ui.zk/WEB-INF/src/org/adempiere/webui/info/InfoWindow.java:739]: // do fill value to editor (for both corrected order and fall back)

File: org.adempiere.ui.zk/WEB-INF/src/org/adempiere/webui/info/InfoWindow.java [hunk 10]
```diff
@@ -617,55 +619,131 @@ public void moveProcessButtonsToBeforeRight() {
...
+				autocompleteEditor.setValue(queryValue);
+				testCount(false);
+			}catch(Exception ex){
+				// => don't run test in case not success set value
+				log.log(Level.SEVERE, "error", ex.getCause());
+			}
+		}else {
+			// => don't run test in case not found auto complete column
+			log.log(Level.SEVERE, String.format("can't found column for autocomplete query for field %s - field id %s. first identify column on m_table need to exists on identifies of info window", 
+					m_gridfield.getColumnName(), m_gridfield.getAD_Column_ID()));
```
Comment (hengsin) [inline org.adempiere.ui.zk/WEB-INF/src/org/adempiere/webui/info/InfoWindow.java:672]: autoCompleteSearchColumn not empty - "Auto complete search column (%s) not found for field %s (field id %s). ", autoCompleteSearchColumn, m_gridfield.getColumnName(), m_gridfield.getAD_Column_ID()

autoCompleteSearchColumn empty - "Info window has no identifier columns"

**Comment by hengsin** [comment]
hi @hieplq , please apply the following patch, thanks.
[pr2029.patch.txt](https://github.com/idempiere/idempiere/files/12777312/pr2029.patch.txt)

**Review by hengsin — APPROVED** [review]
tested

**Pull request merged**